statistics/cmsketch.go

```diff
@@ -44,29 +56,177 @@ func NewCMSketch(d, w int32) *CMSketch {
 	return &CMSketch{depth: d, width: w, table: tbl}
 }
 
+// NewCMSketchWithTopN returns a new CM sketch with TopN elements.
+// total is the size of the whole dataset
+func NewCMSketchWithTopN(d, w int32, data [][]byte, numTop uint32, total uint64) *CMSketch {
+	c := NewCMSketch(d, w)
+	c.BuildTopN(data, numTop, total)
+	return c
+}
+
+func groupElements(data [][]byte) map[hack.MutableString]uint64 {
+	counter := make(map[hack.MutableString]uint64)
+
+	for k := range data {
+		counter[hack.String(data[k])]++
+	}
+
+	return counter
+}
+
+// BuildTopN builds table of top N elements.
+// elements in data should not be modified after this call.
+// If the original data is not skew enough, then we won;t build map for top n elements.
+func (c *CMSketch) BuildTopN(data [][]byte, numTop uint32, total uint64) {
```

Comment:
Can we separate this function into smaller ones follows these steps:
preprocess --> checkEnableTopN --> buildTopN or others